core/src/main/java/io/spine/core/Events.java

```diff
@@ -183,16 +200,80 @@ public static EventId checkValid(EventId id) {
 
     /**
      * Obtains the {@link TenantId} from the given {@link Event}.
+     * 
+     * <p>The {@code TenantId} is retrieved by traversing the passed {@code Event}s context. It is
+     * stored in the initial {@link CommandContext} and can be retrieved from the events origin 
+     * command or rejection context. 
      */
     @Internal
     public static TenantId getTenantId(Event event) {
-        final TenantId result = event.getContext()
-                                     .getCommandContext()
-                                     .getActorContext()
-                                     .getTenantId();
+        checkNotNull(event);
+        
+        final Optional<CommandContext> commandContext = getCommandContext(event);
+
+        if (!commandContext.isPresent()) {
+            return TenantId.getDefaultInstance();
+        }
+
+        final TenantId result = getTenantId(commandContext.get());
         return result;
     }
 
+    /**
+     * Obtains the {@link TenantId} from the {@link CommandContext}.
+     * 
+     * <p>The {@link CommandContext} is accessible from the {@link Event} if the {@code Event} was 
+     * created as a result of some command or its rejection. This makes the {@code CommandContext} 
+     * a valid {@code TenantId} source inside of the {@code Event}.
+     */
+    private static TenantId getTenantId(CommandContext context) {
+        return context.getActorContext()
+                      .getTenantId();
+    }
+
+    /**
+     * Obtains the context of the command, which lead to this event.
+     * 
+     * <p> The context is obtained by traversing the events origin for a valid context source. 
+     * There are can be two sources for the command context:
+     * <ol>
+     *     <li>The command context set as the event origin.</li>
+     *     <li>The command set as a rejection property.</li>
+     * </ol>
+     * 
+     * <p>If at some point the event origin is not set the {@link Optional#absent()} is returned.
+     */
+    private static Optional<CommandContext> getCommandContext(Event event) {
```

Comment:
Assuming the description, the name `getOriginCommandContext()` would highlight the fact that we go deeper.